compiler/src/iree/compiler/Dialect/Flow/Transforms/test/fusion_of_tensor_ops.mlir

```diff
@@ -121,3 +121,42 @@ func.func @batchnorm_training(%10 : tensor<12xf32>, %11 : tensor<12x12x12x12x12x
 //  CHECK-SAME:       ins(%[[ARG0]], %[[GENERIC0]] :
 //  CHECK-SAME:       outs(%[[INIT]], %[[INIT]], %[[INIT]] :
 //       CHECK:   return %[[GENERIC1]]#0, %[[GENERIC1]]#1, %[[GENERIC1]]#2
+
+// -----
+
+#map = affine_map<(d0, d1) -> (d0, d1)>
+module {
+  func.func @fuse_only_with_same_marker(%arg0: tensor<5x5xf32>, %arg1: tensor<5x5xf32>) -> (tensor<5x5xf32>, tensor<5x5xf32>, tensor<5x5xf32>, tensor<5x5xf32>) {
```

Comment:
Not sure how this captured the problem. What was the output without your change?

Reply:
In the first iteration, it will mark `%4 = linalg.generic` to fuse with `%6 = linalg.generic` and `%5 = linalg.generic` to fuse with `%7 = linalg.generic`. Then in the `FuseElementwiseOpsWithMultipleUses`, the consumer `%7 = linalg.generic` will be picked up first. When searching for the producer, its first operand is `%4`. So if we don't check the marker value, it'll fuse with `%4` instead of `%5`. But `%4` shouldn't be fused with `%7` first, because that will violate the dominance of its user `%6` ('%7' is after `%6`).

If we set `multi-use-fusion-iteration=1`, the pass will exit in a state where `%4` is fused to `%7` and `%6` is not dominated, then the validator will find the dominance violation. With this patch, run:
```
iree-opt --pass-pipeline="func.func(iree-flow-fusion-of-tensor-ops{fuse-multi-use=true multi-use-fusion-iteration=1})"
```
on that testcase will trigger the dominance violation.

In this patch (I realized afterward), the test failing without the check on makers is because we assert the producer must exist for each consumer in `FuseElementwiseOpsWithMultipleUses`. If we incorrectly fuse `%4` and `%7`  first, the consumer '%6' won't be able to find the producer and fail the assertion. It doesn't directly test the dominance violation I just mentioned, but still shows the violation on another assumption. I can update the test if you think we should test the dominance violation.